doc/config.md

```diff
@@ -264,6 +264,27 @@ updated and how often aggregated statistics are written to the log
 
 Default: 60
 
+### disable_prepared_statement_support
+
+Enables/disables server side prepared statement support. Disabled by
+default. To enable the server-side prepared statement support change
+it to 0 (zero).
+
+Default: 1 (disabled)
+
+### prepared_statement_cache_queries
```

Comment:
Done.
I also renamed `prepared_statement_cache_size` to 'prepared_statement_cache_size`
because "statement" ... "queries" seems to be duplication.

Reply:
Maybe better to disable it by default? Seems the default cache size is 100 and the code is now also executed in session and statement pooling mode.

You might take a look at [this](https://github.com/topicusonderwijs/pgbouncer-ps-patch/commit/a4c770d5357e9a2d7bd09ff3fe802fdba93f8f69) commit for some additional inspiration. Basically the 'enabled' check is as follows:
```
(pool_pool_mode(pool) == POOL_TX && pool_prepared_statement_cache_queries(pool) > 0)
```
The cache size setting has been duplicated on database and user level, in addition to global level. Allowing more fine grained enabling of prepared statement support on the pool level.

Reply:
Supporting prepared statements only in transaction mode definitely makes sense.
I have changed it.
But I am not sure that 
1. We should switch off support of prepared statements by default (`prepared_statements_cache_size=0`): in session mode them are switched off in any case and if client tries to use prepared statements in transactional mode without enable support of them in pgbouncer then h will get error, if prepared statements are not user by clients, then there is no difference whether prepared statements support is enable or not.
2. It makes sense to specify prepared statement cache size per database or user. Why do we need to limit number of cached prepard statements for one database by lets say 100 and for another - 1000? It is not a problem is actual number of used prepared statements is smaller than specified cache size - just this cache will be partly filled. Fro the other size if some of actively used prepared statements are evicted from the cache, then we will get trashing, with negative impact on performance.

Reply:
Some "prepared statement" code paths do get executed in transaction mode, even if the client is not using prepared statements. E.g. a client sending a 'Describe' message for a portal. We do have to parse the message to determine if it is of type 'S' and we should handle it. Apart from performance it also reduces the risk of "prepared statement" code causing issues for users not even using prepared statements. The latter is/was my main reason to put everything behind an on/off switch.

Keeping prepared statements also have a performance impact on the postgresql server itself. I don't know how everyone is running their pgbouncer setup, so it is just giving users the freedom to configure it for their setup.

Reply:
Mostly agree with you.
Change default value of `prepared_statements_cache_size` to 0 (disable)
But still I found out it very bad and confusing when people are trying to use prepared statements with enabled connection pooler and got errors. What is worse - everything can work for some time until number of connection doesn't exceed number of workers an then you will get error.

Do you think it will be important to support describe portal message if support of prepared statements is enabled? I wonder who ever used this command and why?

Reply:
_I didn't review the current patch yet but I spent some time on this topic._ This current feature is related to support prepared statements in transaction pooling mode. It doesn't make sense to disable prepared statements in session pooling mode (that's always on) or to allow enabling prepared statements in statement pooling mode (is there any benefit? I expect some performance regression.)
Regarding the default setting for this parameter, it will depend on the change behavior. If it doesn't break scenarios, there is no performance regression and we are confident that this feature is mature, we can consider enable it by default. If there are concerns about it, it should be disabled by default and we can evaluate changing this setting later.

Reply:
This is the current behavior: by default `prepared_statements_cache_size=0` which disable prepared statement support by pgbouncer in transaction mode. But it doesn't mean that we prohibit to use prepared statements in this or any other mode. Behavior will be the same as with master version of pgbouncer.